src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,260 @@
+            await handler.RunAsync(stage);
+            if (handler.ExecutionContext.Result == TaskResult.Failed)
+            {
+                throw new Exception($"The hook script at '{HookIndexPath}' running command '{input.Command}' did not execute successfully."); // TODO: better exception
+            }
+            var response = GetResponse(input);
+            return response;
+        }
+
+        private string GenerateResponsePath()
+        {
+            return Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Temp), ResponseFolderName, $"{Guid.NewGuid()}.json");
+        }
+
+        private string GetPrependPath(IExecutionContext context)
+        {
+            return string.Join(Path.PathSeparator.ToString(), context.Global.PrependPath.Reverse<string>());;
+        }
+
+        private HookResponse GetResponse(HookInput input)
+        {
+            HookResponse response = null;
+
+            if (!string.IsNullOrEmpty(input.ResponseFile) && File.Exists(input.ResponseFile))
+            {
+                response = IOUtil.LoadObject<HookResponse>(input.ResponseFile);
+                IOUtil.DeleteFile(input.ResponseFile);
+                Trace.Info($"Response file for the hook script at '{HookIndexPath}' running command '{input.Command}' successfully processed and deleted.");
+
+                // IsAlpine is mandatory for prepare_job hook
+                if (input.Command == HookCommand.PrepareJob && response.IsAlpine == null)
+                {
+                    throw new Exception("The property 'isAlpine' is required but was not found in the response file.");
+                }
+            }
+            else
+            {
+                Trace.Info($"Response file for the hook script at '{HookIndexPath}' running command '{input.Command}' not found.");
+                if (input.Command == HookCommand.PrepareJob)
+                {
+                    throw new Exception($"Response file is required but not found for the hook script at '{HookIndexPath}' running command '{input.Command}'");
+                }
+            }
+
+            return response;
+        }
+
+        private static JToken GetHookStateInJson(IExecutionContext context)
+        {
+            if (context.JobContext.TryGetValue("hook_state", out var hookState))
+            {
+                return JsonUtility.FromString<JToken>(hookState.ToString());
+            }
+            return null;
+        }
+
+        private void SaveHookState(IExecutionContext context, JToken hookState, HookInput input)
+        {
+            // TODO: consider JTokenContextData
+            if (hookState == null)
+            {
+                Trace.Info($"No 'state' property found in response file for '{input.Command}'. Context variable 'hook_state' will not be updated.");
+                return;
+            }
+            context.JobContext["hook_state"] = new StringContextData(JsonUtility.ToString(hookState));
```

Comment:
is this going to show on the `job` context? do we want that?

Reply:
Should we make a conscious effort to hide it?

It does allow the hooks to make arbitrary information, **potentially unregistered secrets** available for printing in the workflow. Someone unfamiliar with the implementation would not expect this.

Any other concerns? Otherwise, I see how hiding `hook_state` from the job context would be a good 'least privilege / access'  type of improvement.

Reply:
I think we should be very deliberate about what we have in the context, as that is an interface between our system and workflow authors. Unless we have a reason to put it in there, I don't think we should. We can always add it in the future based on feedback.

Reply:
What do you think of putting it in `ExecutionContext.Global`? We use `GlobalContext` for a lot of non-customer facing info, like steps and job telemetries.